coreneuron/io/lfp.cpp

```diff
@@ -0,0 +1,139 @@
+                throw std::invalid_argument(s.str());
+            }
+            return std::abs(std::log(b / a)) / dxn;
+        } else {
+            return std::log((b + std::sqrt(b * b + q2)) / (a + std::sqrt(a * a + q2))) / dxn;
+        }
+    };
+    if (delta <= 0.0) {
+        return f * log_integral(-mu, one_m_mu);
+    } else {
+        double sqr_delta(std::sqrt(delta));
+        double d1(mu - sqr_delta);
+        double d2(mu + sqr_delta);
+        double parts = 0.0;
+        if (d1 > 0.0) {
+            double b(std::min(d1, 1.0) - mu);
+            parts += log_integral(-mu, b);
+        }
+        if (d2 < 1.0) {
+            double b(std::max(d2, 0.0) - mu);
+            parts += log_integral(b, one_m_mu);
+        };
+        // complement
+        double maxd1_0(std::max(d1, 0.0)), mind2_1(std::min(d2, 1.0));
+        if (maxd1_0 < mind2_1) {
+            parts += 1.0 / radius * (mind2_1 - maxd1_0);
+        }
+        return f * parts;
+    };
+}
+}  // namespace lfputils
+
+using namespace lfputils;
+
+template <LFPCalculatorType Type, typename SegmentIdTy>
+LFPCalculator<Type, SegmentIdTy>::LFPCalculator(const Point3Ds& seg_start,
+                                                const Point3Ds& seg_end,
+                                                const std::vector<double>& radius,
+                                                const std::vector<SegmentIdTy>& segment_ids,
+                                                const Point3Ds& electrodes,
+                                                double extra_cellular_conductivity)
+    : segment_ids_(segment_ids) {
+    if (seg_start.size() != seg_end.size()) {
+        throw std::invalid_argument("Different number of segment starts and ends.");
+    }
+    if (seg_start.size() != radius.size()) {
+        throw std::invalid_argument("Different number of segments and radii.");
+    }
+    double f(1.0 / (extra_cellular_conductivity * 4.0 * 3.141592653589));
+
+    m.resize(electrodes.size());
+    for (size_t k = 0; k < electrodes.size(); ++k) {
+        auto& ms = m[k];
+        ms.resize(seg_start.size());
+        for (size_t l = 0; l < seg_start.size(); l++) {
+            ms[l] = getFactor(electrodes[k], seg_start[l], seg_end[l], radius[l], f);
+        }
+    }
+}
+
+template <LFPCalculatorType Type, typename SegmentIdTy>
+template <typename Vector>
+inline void LFPCalculator<Type, SegmentIdTy>::lfp(const Vector& membrane_current) {
+    std::vector<double> res(m.size());
+    for (size_t k = 0; k < m.size(); ++k) {
+        res[k] = 0.0;
+        auto& ms = m[k];
+        for (size_t l = 0; l < ms.size(); l++) {
+            res[k] += ms[l] * membrane_current[segment_ids_[l]];
+        }
+    }
+    std::vector<double> res_reduced(res.size());
+#if NRNMPI
+    int mpi_sum{1};
+    nrnmpi_dbl_allreduce_vec(res.data(), res_reduced.data(), res.size(), mpi_sum);
+#else
+    res_reduced = res;
+#endif
+    lfp_values = res_reduced;
```

Comment:
This chunk of code makes a lot of unnecessary vector allocations and copies. I do not understand why variable `res_reduced` is necessary, and we should use `std::swap` when not using MPI instead of doing a vector copy, something like below should work:

```diff
diff --git a/coreneuron/io/lfp.cpp b/coreneuron/io/lfp.cpp
index 00716fd..f4812c1 100644
--- a/coreneuron/io/lfp.cpp
+++ b/coreneuron/io/lfp.cpp
@@ -108,14 +108,14 @@ inline void LFPCalculator<Type, SegmentIdTy>::lfp(const Vector& membrane_current
             res[k] += ms[l] * membrane_current[segment_ids_[l]];
         }
     }
-    std::vector<double> res_reduced(res.size());
+
 #if NRNMPI
     int mpi_sum{1};
-    nrnmpi_dbl_allreduce_vec(res.data(), res_reduced.data(), res.size(), mpi_sum);
+    lfp_values.resize(res.size());
+    nrnmpi_dbl_allreduce_vec(res.data(), lfp_values.data(), res.size(), mpi_sum);
 #else
-    res_reduced = res;
+    std::swap(res, lfp_values);
 #endif
-    lfp_values = res_reduced;
 }
```

Also the `res` variable is not necessary. `lfp_values` should be used directly instead. (assuming that `nrnmpi_dbl_allreduce_vec` supports in-place operation`. @pramodk ?


In term of design, it seems to be the LFPCalculator class should not have a `lfp_values` member variable. IMO the `lfp` member method should be const and returns the result directly. (The class can also be a functor by renaming the `lfp` member function by `std::vector<double> operator()(const Vector& membrane_current) const` function.

Reply:
Ok. res variable is removed. 

lfp values need to be cached somehow within the simulation and read subsequently. 
It could be done elsewhere. I would think that for now it is ok to store them within the simulator.